# postmaster

Postmaster is a tool for annotating transcriptome alignments with posterior 
alignment probabilities derived from [salmon](https://github.com/COMBINE-lab/salmon)
quantifications.  In short, it takes as input a SAM or BAM file generated by 
salmon using the `--writeMappings` option, and outputs an equivalent 
SAM or BAM file where each alignment is additionally annotated with a `ZW` 
attribute (the same attribute used by [RSEM](https://github.com/deweylab/RSEM))
denoting the posterior probability of this particular alignment.

## Usage

The required input to run `postmaster` is an alignment file (generated by 
salmon with the `--writeMappings` option) and a `quant.sf` file from the 
corresponding run of salmon.  If no explicit `-a` is provided, then 
`postmaster` will assume it is to consume input from `stdin`.  The output
(annotated) alignments will be written to the file given to the `-o` option,
and the format will be determined by the filename's suffix â€” if `.sam` then
a SAM format file will be written and if `.bam` then a BAM format file will
be written.  If no `-o` option is provided, then `postmaster` will instead
write the output to `stdout` **in BAM format**.

```
USAGE:
postmaster [OPTIONS] --quant <QUANT>

OPTIONS:
-a, --alignments <ALIGNMENTS>
-h, --help                         Print help information
-n, --num-threads <NUM_THREADS>    [default: 2]
-o, --output <OUTPUT>
-q, --quant <QUANT>
-V, --version                      Print version information
```